src/nemos/basis.py

```diff
@@ -845,127 +886,136 @@ def evaluate_on_grid(self, n_samples: int) -> Tuple[NDArray, NDArray]:
-
-        Map the equi-spaced samples from [0,1] to log equi-spaced samples [0, (n_basis_funcs - 1) * pi].
+        """
+        Map the sample domain to log-space.
 
         Parameters
         ----------
-        sample_pts :
-            The sample points used for evaluating the splines, shape (n_samples,).
+        sample_pts : NDArray
+            Sample points used for evaluating the splines,
+            shape (n_samples, ).
 
         Returns
         -------
-        :
-            A transformed version of the sample points that matches the Raised Cosine basis domain,
-            shape (n_sample_points, ).
+        NDArray
+            Transformed version of the sample points that matches the Raised Cosine basis domain,
+            shape (n_samples, ).
         """
-        return (
-            np.power(
-                10,
-                -(np.log10((self.n_basis_funcs - 1) * np.pi) + 1) * sample_pts
-                + np.log10((self.n_basis_funcs - 1) * np.pi),
-            )
-            - 0.1
+        # This log-stretching of the sample axis has the following effect:
+        # - as the time_scaling tends to 0, the points will be linearly spaced.
+        # - as the time_scaling tends to inf, basis will be dense around 0.
+        log_spaced_pts = np.log(self.time_scaling * sample_pts + 1) / np.log(
+            self.time_scaling + 1
         )
+        return log_spaced_pts
 
-    def _check_n_basis_min(self) -> None:
-        """Check that the user required enough basis elements.
+    def evaluate(self, sample_pts: ArrayLike) -> NDArray:
+        """Generate log-spaced raised cosine basis with given samples.
+
+        Parameters
+        ----------
+        sample_pts :
+            Spacing for basis functions, holding elements on interval [0, 1].
 
-        Checks that the number of basis is at least 2.
+        Returns
+        -------
+        basis_funcs :
+            Log-raised cosine basis functions, shape (n_samples, n_basis_funcs).
 
         Raises
         ------
         ValueError
-            If an insufficient number of basis element is requested for the basis type
+            If the sample provided do not lie in [0,1].
         """
-        if self.n_basis_funcs < 2:
-            raise ValueError(
-                f"Object class {self.__class__.__name__} requires >= 2 basis elements. "
-                f"{self.n_basis_funcs} basis elements specified instead"
-            )
+        (sample_pts,) = self._check_evaluate_input(sample_pts)
+        if not self.enforce_decay_to_zero:
+            # keep the last basis, i.e. do not enforce decay to zero
+            # for the filter.
+            eval_basis = super().evaluate(self._transform_samples(sample_pts))
+        else:
+            # temporarily add n_trim basis element
+            # guaranteeing that the last basis element decays to 0 for any width.
+            n_trim = int(np.ceil(self.width))
+            self.n_basis_funcs += n_trim
```

Comment:
this kind of modification feels hacky. An alternative would be to have `evaluate` accept `n_basis_funcs` as an optional arg, defaulting to `self.n_basis_funcs`. But we don't want the public method to accept that argument, so not sure...

Reply:
I agree, the other way would be to be smarter in the way we compute the stretching. the equivalent operation would be to modify the scaling on the fly, but leaving the normalization constant (log(self.time_scaling + 1)) as it is.
basically the transform samples should look something like:
```python
    def _transform_samples(self, sample_pts: NDArray, scaling: float) -> NDArray:
        """
        Map the sample domain to log-space.

        Parameters
        ----------
        sample_pts :
            Sample points used for evaluating the splines,
            shape (n_samples, ).

        Returns
        -------
            Transformed version of the sample points that matches the Raised Cosine basis domain,
            shape (n_samples, ).
        """
        # This log-stretching of the sample axis has the following effect:
        # - as the time_scaling tends to 0, the points will be linearly spaced.
        # - as the time_scaling tends to inf, basis will be dense around 0.
        log_spaced_pts = np.log(scaling * sample_pts + 1) / np.log(
            self.time_scaling + 1
        )
        return log_spaced_pts
```

And the scaling will be equal to self.time_scaling, if no decay to zero is needed, some other constant computed appropriately otherwise.
what do you think of this solution?

Reply:
I think this solution is fine but it's not clear to me why this is equivalent. And I worry about being able to explain it clearly and succinctly.

The other alternative would be to have a private helper method `_shift_sample_pts`, which accepts `n_basis_funcs` and runs:
```python
        peaks = np.linspace(0, 1, self.n_basis_funcs)
        delta = peaks[1] - peaks[0]
        # generate a set of shifted sample points between 0 and 1
        return sample_pts[:, None] - peaks[None]) / (delta * self.width)
```

Or could potentially include the multiplication by pi and the clipping.

Or clip between -1 and 1 (and then `evaluate` just has the multiplication by pi, no clipping).

Reply:
And evaluate checks `self.enforce_decay_to_zero` to determine `n_basis_funcs`?

Reply:
in order to achieve the desired effect with the shifting, we should have that for the last basis that the peaks do not terminate in 1:

```
def scale_pts(time_scaling, width, n_basis_funcs, sample_pts):
    log_spaced_pts = np.log(time_scaling * sample_pts + 1) / np.log(
        time_scaling + 1
    )
    peaks = np.linspace(0, 1 - np.ceil(width) / (n_basis_funcs + np.ceil(width) - 1), n_basis_funcs)
    delta = 1 / (n_basis_funcs+ np.ceil(width) -1)
    basis_funcs = 0.5 * (
            np.cos(
                np.clip(
                    np.pi * (log_spaced_pts[:, None] - peaks[None]) / (delta * width),
                    -np.pi,
                    np.pi,
                )
            )
            + 1
    )
    return basis_funcs
hacky_version = nmo.basis.RaisedCosineBasisLog(10,width=2, time_scaling=20).evaluate_on_grid(1000)[1]
adjusting_peaks = scale_pts(time_scaling=20., width=2, n_basis_funcs=10, sample_pts=np.linspace(0,1,1000))
print(np.linalg.norm(hacky_version-adjusting_peaks))
```

Reply:
the function above compute the peaks appropriately, the delta is adjusted to maintaining the original spacing. The same result could be achieved by adjusting the log stretching, as I mentioned previously.

Reply:
My point is that, in order to have the math work out exactly, we need some computation that is not exactly transparent; and i think we need to preserve the transformation that we have now (the hacky solution of trimming basis). 

in the sense that the hacky solution is the right thing to do mathematically because we want the same temporal log-stretching for a given time_scaling parameter. if we simply adjust the peak locations and not the delta at the denominator of the raised cosines, this would not be true anymore.

Reply:
I computed the peaks conditionally on the enforce_decay_to_zero flag, not it is easy to understand what's happening. the last peak location ensures that the last basis element decays to zero

Reply:
I think the current solution makes sense.